src/main/java/dev/slimevr/vr/trackers/udp/UDPDevice.java

```diff
@@ -27,6 +28,7 @@ public class UDPDevice extends Device {
 	public int firmwareBuild = 0;
 	public boolean timedOut = false;
 	private final FastList<Tracker> trackers = new FastList<>();
+	private final ConcurrentHashMap<Integer, Integer> remote_id_map = new ConcurrentHashMap<>();
```

Comment:
please change it to `HashMap<Integer, Tracker> trackerByRemoteId = new HashMap<>();` there is no need to use additional re-mapping of IDs you can fetch trackers directly, and there is no need to use ConcurrentHashMap as all UDP processing is done in one thread.

Reply:
> there is no need to use additional re-mapping of IDs you can fetch trackers directly

What do you mean with this? Tracker sensor ID's must be created in order, starting with 0, going 1, 2, 3 etc.
The row above you see is a FastList, which is not able to have arbitrary ID's. This is a problem when UDP packets are sent in the wrong order. My trackers sometimes send a set up packet for sensor 1, then sensor 0. This means that the sensor 1 info is set to the sensor 0 info, and the sensor 0 info is ignored, because it "already exists".

Reply:
You don't need to map remote sensor id to local sensor id, you can map remote sensor id directly to the sensor object. Also the name of the variable is wrong, java uses camelCase.

Reply:
Other code needs the trackers object to be a list that is contiguous. A bigger rewrite of the TrackersUDPServer.java file might be nice too, but right now this is the simplest way to add a mapping layer.

Reply:
You can literally have both, you know.

Reply:
I understand the problem, I just think that it's not a good solution to it. It's the one with the least code change, yes, but it's not the best way. Either you need to duplicate the list and add a map nearby, or replace the list with the map and change the code that uses the list, imo.